### PR TITLE
Add `--dry-run` flag to `uv pip install`

### LINKED ISSUE #1244: Add `--dry-run` flag for `pip install`
For parity with `pip` but also useful.

Basically this let's you `pip compile` without creating requirements file. The install plan is displayed to the user, e.g.

```
❯ pip install flask --dry-run
Collecting flask
  Using cached flask-3.0.1-py3-none-any.whl (101 kB)
Requirement already satisfied: Werkzeug>=3.0.0 in /Users/mz/.pyenv/versions/3.11.4/lib/python3.11/site-packages (from flask) (3.0.1)
Requirement already satisfied: Jinja2>=3.1.2 in /Users/mz/.pyenv/versions/3.11.4/lib/python3.11/site-packages (from flask) (3.1.2)
Requirement already satisfied: itsdangerous>=2.1.2 in /Users/mz/.pyenv/versions/3.11.4/lib/python3.11/site-packages (from flask) (2.1.2)
Requirement already satisfied: click>=8.1.3 in /Users/mz/.pyenv/versions/3.11.4/lib/python3.11/site-packages (from flask) (8.1.7)
Requirement already satisfied: blinker>=1.6.2 in /Users/mz/.pyenv/versions/3.11.4/lib/python3.11/site-packages (from flask) (1.7.0)
Requirement already satisfied: MarkupSafe>=2.0 in /Users/mz/.pyenv/versions/3.11.4/lib/python3.11/site-packages (from Jinja2>=3.1.2->flask) (2.1.3)
Would install Flask-3.0.1
```

### PR DESCRIPTION
**Pull request by JacobCoffee**
## What

Adds a `--dry-run` flag that ejects out of the installation process early (but after resolution) and displays only what *would have* installed

## Closes

Closes #1244 

## Out of Scope

I think it may be nice to include a `dry-run` flag for `uninstall` even though `pip` doesn't implement this... thinking `Would uninstall X packages: ...`

### CODE REVIEW

**Comment by JacobCoffee** [comment]
image

```
➜ ./target/debug/uv pip install litestar --dry-run
Resolved 22 packages in 584ms
Would install 22
 - msgspec==0.18.6
 - rich-click==1.7.3
 - pyyaml==6.0.1
 - python-dateutil==2.8.2
 - httpcore==1.0.3
 - polyfactory==2.14.1
 - httpx==0.26.0
 - click==8.1.7
 - certifi==2024.2.2
 - anyio==4.2.0
 - h11==0.14.0
 - six==1.16.0
 - sniffio==1.3.0
 - rich==13.7.0
 - faker==23.2.0
 - litestar==2.6.1
 - mdurl==0.1.2
 - pygments==2.17.2
 - typing-extensions==4.9.0
 - multidict==6.0.5
 - idna==3.6
 - markdown-it-py==3.0.0
```

**Comment by JacobCoffee** [comment]
Sorry if this is a bad way to get the version? It felt bad, anyway...

**Comment by JacobCoffee** [comment]
Might be cool to have a `dry-run` flag for uninstalling although not native to `pip`

File: crates/uv/src/commands/pip_install.rs [hunk 1]
```diff
@@ -237,6 +238,35 @@ pub(crate) async fn pip_install(
...
+        for package_name in resolution.packages() {
+            if let Some(dist) = resolution.get(package_name) {
+                let version = dist
+                    .version()
+                    .map_or_else(String::new, |version| format!("=={version}"));
+
+                writeln!(
+                    printer,
+                    " {} {}{}",
+                    "~".blue(),
```
Comment (JacobCoffee) [inline crates/uv/src/commands/pip_install.rs:261]:
```suggestion
                    "+".blue(),
```
Not sure how this should look? `+` and `-` didn't seem to fit, though.. so I tried `~`

**Comment by JacobCoffee** [comment]
On the failing CI, I had some failures do to order in stdout there... but I'm wondering how we keep it consistent?

**Comment by zanieb** [comment]
Hey @JacobCoffee, thanks for the pull request! I think we'll actually want to implement this a bit differently e.g. by passing the dry run flag into `install`

https://github.com/astral-sh/uv/blob/896ab1c54f4f4052555fa910d523f6899685eed2/crates/uv/src/commands/pip_install.rs#L477

then we can use our existing display (which has a sorting implementation for deterministic output)

https://github.com/astral-sh/uv/blob/896ab1c54f4f4052555fa910d523f6899685eed2/crates/uv/src/commands/pip_install.rs#L642

We'll just need to tweak the `install` function to avoid doing any downloads, uninstalls, builds, or installs when the flag is set.

**Comment by JacobCoffee** [comment]
Does this look okay? also not 100% sure how to trigger re-downloads to test that bit.
Also, if it works out I can write the tests for it 

```
uv on  unst [📝🤷] via 🐋 colima via 🎁 v0.1.1 via  pyenv via ⚙️ v1.75.0
➜ ./target/debug/uv pip install litestar --dry-run
Resolved 22 packages in 48ms
DRY RUN The following packages would be installed from local cache:
 ~ litestar==2.6.1

uv on  unst [📝🤷] via 🐋 colima via 🎁 v0.1.1 via  pyenv via ⚙️ v1.75.0
➜ ./target/debug/uv pip install litestar --dry-run -n
Resolved 22 packages in 3.60s
DRY RUN The following packages would be downloaded:
 ~ litestar==2.6.1
```

image

**Comment by zanieb** [comment]
I'll give this a try tomorrow!

**Comment by zanieb** [comment]
Hey! I got a chance to play with this today. I'm thinking we should match our normal output much more closely.

Screenshot 2024-02-22 at 18 27 31

I put up a commit for you to build off of at https://github.com/astral-sh/uv/pull/1890

We still need more test coverage i.e. install a dependency of a package first then dry run install what do we show? or install an old version then dry run install with `--upgrade`

**Comment by JacobCoffee** [comment]
Just for my notes, i am running into an issue where in testing it passes with 2 `$package==$version`, but in CLI usage i get 4 `====`
```
➜ ./target/debug/uv pip install litestar==2.0.0 --dry-run --strict
Resolved 17 packages in 974ms
Would download 2 packages
Would install 17 packages
 + anyio====4.3.0
 + litestar====2.0.0
 + faker====23.2.1
 + fast-query-parsers==1.0.3
```
But removing the `==` printer config from the `+`/`-` sections remove it from the testing :|

**Comment by zanieb** [comment]
@JacobCoffee I'm on vacation this week but let me know when this is ready for review again

**Comment by JacobCoffee** [comment]
sorry ive not wrapped this - i think my remaining issue is the 4 ====, but ive just had surgery so im down for a bit on my hard thinking 😅

**Comment by zanieb** [comment]
I can give it a look! Get better :)

**Comment by zanieb** [comment]
The issue is that `InstalledVersion` which comes from cached distributions includes `==` already (fixed by https://github.com/astral-sh/uv/pull/1436/commits/fee46b34aafd1fe400f38fa7ed0517885e2a332b).

File: crates/uv/tests/pip_install.rs [hunk 2]
```diff
@@ -2247,3 +2247,219 @@ requires-python = ">=3.11,<3.13"
...
+    exit_code: 101
+    ----- stdout -----
+
+    ----- stderr -----
+    Resolved 3 packages in [TIME]
+    Would download 3 packages
+    Would install 3 packages
+    thread 'main' panicked at crates/uv/src/commands/pip_install.rs:785:65:
+    called `Option::unwrap()` on a `None` value
+    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Comment (zanieb) [inline crates/uv/tests/pip_install.rs:2306]: Need to avoid panicking here

**Comment by zanieb** [comment]
I think this is vaguely ready for review

File: crates/uv/src/commands/pip_install.rs [hunk 3]
```diff
@@ -704,6 +717,135 @@ async fn install(
...
+                format!(
+                    "Audited {} in {}",
+                    format!("{} package{}", resolution.len(), s).bold(),
+                    elapsed(start.elapsed())
+                )
+                .dimmed()
+            )?;
+            writeln!(printer, "Would make no changes")?;
+            return Ok(());
+        }
```
Comment (zanieb) [inline crates/uv/src/commands/pip_install.rs:749]: I believe this is redundant as it is handled before this function is called

File: crates/uv/src/commands/pip_install.rs [hunk 4]
```diff
@@ -704,6 +717,135 @@ async fn install(
...
+
+        // Install the resolved distributions.
+        let installs = wheels.len() + local.len();
+
+        if installs > 0 {
+            let s = if installs == 1 { "" } else { "s" };
+            writeln!(
+                printer,
+                "{}",
+                format!("Would install {}", format!("{installs} package{s}").bold(),).dimmed()
```
Comment (zanieb) [inline crates/uv/src/commands/pip_install.rs:801]: Formatting needs fix

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**